### PR TITLE
[port] Update BU random number generator

### PR DESCRIPTION
**Pull request by sickpig**
This PR update BU code with the latest Core's when it comes to random number generator especially for what we use in C++ unit tests,but also in other part of the code base.

This the list of Core PR included in this port:

- bitcoin/bitcoin/pull/9821
- bitcoin/bitcoin/pull/9792
- bitcoin/bitcoin/pull/7891
- bitcoin/bitcoin/pull/10372
- bitcoin/bitcoin/pull/10321
- bitcoin/bitcoin/pull/9480
- bitcoin/bitcoin/pull/9349
- bitcoin/bitcoin/pull/10377
- bitcoin/bitcoin/pull/10837
- bitcoin/bitcoin/pull/10820
- bitcoin/bitcoin/pull/10855
- bitcoin/bitcoin/pull/13314
- bitcoin/bitcoin/pull/12724
- bitcoin/bitcoin/pull/14624

An alternative to this approach will be to take the zcash path suggested by @dagurval and that has been partially ported over XT in this wip PR

### CODE REVIEW

File: src/init.cpp [hunk 1]
```diff
@@ -1554,8 +1562,6 @@ bool AppInit2(Config &config, thread_group &threadGroup)
     if (!strErrors.str().empty())
         return InitError(strErrors.str());
 
-    RandAddSeedPerfmon();
-
```
Comment (ptschip) [inline src/init.cpp:1551]: maybe a dumb question but why take this out , don't we still need something here for a seed?

File: src/prevector.h [hunk 2]
```diff
@@ -360,6 +360,7 @@ class prevector
         }
     }
 
+    prevector(prevector<N, T, Size, Diff> &&other) : _size(0) { swap(other); }
     prevector &operator=(const prevector<N, T, Size, Diff> &other)
```
Comment (ptschip) [inline src/prevector.h:364]: what do these prevector changes have to do with random numbers or did they just slip in during cherry-picking?

**Comment by dagurval** [comment]
> An alternative to this approach will be to take the zcash path suggested by @dagurval and that has been partially ported over XT

I don’t think it’s worth taking the libsodium approach now that you’ve done all this cheery-picking work!

File: src/prevector.h (hunk 2 above)
Comment (sickpig) [reply src/prevector.h:364]: This come from bitcoin/bitcoin/pull/9349 changes `src/test/prevectors_test.cpp`, some file has been modified by the proper random cherry pick I made, so I decided to port also this PR. 

In hindsight I could have left it out cause it is not related to the focus of this PR.

File: src/init.cpp (hunk 1 above)
Comment (sickpig) [reply src/init.cpp:1551]: This change has been ported from bitcoin/bitcoin/pull/7891.

The seed initialization has been moved closer to place where we actually use it, i.e. mainly private key generation.

To go in more detail  `RandAddSeedPerfmon()` ahs been included in `GetStrongRandBytes()` which is the function delegated to initialize the seed for the RNG. At the end of the port this initialization will use 3 sources: OpenSSL RNG, OS RNG and HW RNG (if available).

**Comment by sickpig** [comment]
guess we have some problem while cross compiling for win32/64, or better it seems to be a linker error on libbitcoincash (cashlib):

```
  CXXLD    libbitcoincash.la
.libs/libbitcoincash_la-random.o:random.cpp:(.text+0x2d): undefined reference to `__imp_pthread_mutex_destroy'
.libs/libbitcoincash_la-random.o:random.cpp:(.text+0xba3): undefined reference to `__imp_nanosleep'
.libs/libbitcoincash_la-random.o:random.cpp:(.text+0xea7): undefined reference to `__imp_pthread_mutex_lock'
.libs/libbitcoincash_la-random.o:random.cpp:(.text+0xf8f): undefined reference to `__imp_pthread_mutex_unlock'
.libs/libbitcoincash_la-random.o:random.cpp:(.text+0x1dad): undefined reference to `__imp_pthread_mutex_lock'
.libs/libbitcoincash_la-random.o:random.cpp:(.text+0x1e7c): undefined reference to `__imp_pthread_mutex_unlock'
.libs/libbitcoincash_la-random.o:random.cpp:(.text+0x1eeb): undefined reference to `__imp_nanosleep'
.libs/libbitcoincash_la-random.o:random.cpp:(.text$_ZNSt11unique_lockISt5mutexE6unlockEv[_ZNSt11unique_lockISt5mutexE6unlockEv]+0x2a): undefined reference to `__imp_pthread_mutex_unlock'
.libs/libbitcoincash_la-random.o:random.cpp:(.text.startup+0x4b): undefined reference to `__imp_pthread_mutex_init'
collect2: error: ld returned 1 exit status
Makefile:4081: recipe for target 'libbitcoincash.la' failed
make[3]: *** [libbitcoincash.la] Error 1
make[3]: *** Waiting for unfinished jobs....
```

**Comment by ptschip** [comment]
needs rebase

**Pull request merged**